torch/export/dynamic_shapes.py

```diff
@@ -612,22 +617,85 @@ def is_leaf(t):
         # as well as user-defined classes registered with pytree, which are.
         return _get_node_type(t) not in BUILTIN_TYPES
 
-    def f(t, *dynamic_shapes):
+    def f(path, t, *dynamic_shapes):
         typ = _get_node_type(t)
         # typ is not in BUILTIN_TYPES
         if typ in SUPPORTED_NODES:
             # thus typ is a user-defined class registered with pytree,
             # in which case flatten and recurse
-            return tree_map(
+            return tree_map_with_path(
                 f,
                 SUPPORTED_NODES[typ].flatten_fn(t)[0],
                 *dynamic_shapes,
                 is_leaf=is_leaf,
             )
         else:
-            return func(t, *dynamic_shapes)
+            return func(path, t, *dynamic_shapes)
 
-    return tree_map(f, tree, *dynamic_shapes, is_leaf=is_leaf)
+    try:
+        return tree_map_with_path(f, tree, *dynamic_shapes, is_leaf=is_leaf)
+    except ValueError as e:
+        if "mismatch" in e.args[0]:
+            # When PyTree finds a structural mismatch between tree and dynamic_shapes,
+            # the error message is unfortunately quite horrible. Let's fix that.
+            assert dynamic_shapes, "Cannot be a mismatch if there is no dynamic_shapes"
+            assert tree_name, "Must provide a tree_name when there might be a mismatch"
+
+            def _key(type_, context, i):
+                # derive a PyTree key given the type, context, and child # of a TreeSpec
+                if type_ is dict:
+                    return MappingKey(context[i])
+                if type_ in (list, tuple):
+                    assert context is None
+                    return SequenceKey(i)
+                raise AssertionError(f"Did not expect type {type_}")
```

Comment:
FYI, should look into 'f{value=}' f-string formatter. Can give informative error messages with a nice shorthand.